File: tools/os-test/verify_network_idle_rigorous.py
import json
import time
import urllib.request
import urllib.parse
import threading
import http.server
import socketserver
import socket
from pathlib import Path

BASE_URL = "http://127.0.0.1:58261"

# HTML for testing Network Idle
# It fetches multiple resources including an image and CSS.
TEST_PAGE_HTML = """
<!DOCTYPE html>
<html>
<head>
    <title>Network Idle Test</title>
    <link rel="stylesheet" href="/slow-css">
</head>
<body>
    <h1>Network Idle Test</h1>
    <img src="/slow-image">
    <div id="logs"></div>
    <script>
        // Start fetch immediately, before anything else
        fetch('/slow-resource-1').then(() => {
            console.log("Fetch 1 complete.");
            document.body.innerHTML += "<p id='done'>Done</p>";
        });
    </script>
</body>
</html>
"""

class SlowHandler(http.server.SimpleHTTPRequestHandler):
    def log_message(self, format, *args):
        # Suppress logging to keep output clean
        pass

    def do_GET(self):
        if self.path == "/":
            self.send_response(200)
            self.send_header("Content-type", "text/html")
            self.end_headers()
            self.wfile.write(TEST_PAGE_HTML.encode())
        elif self.path == "/slow-css":
            time.sleep(1)  # Simulate slow CSS
            self.send_response(200)
            self.send_header("Content-type", "text/css")
            self.end_headers()
            self.wfile.write(b"body { color: red; }")
        elif self.path == "/slow-image":
            time.sleep(2)  # Simulate slow image
            self.send_response(200)
            self.send_header("Content-type", "image/png")
            self.end_headers()
            # Minimal 1x1 transparent PNG
            self.wfile.write(b'\\x89PNG\\r\\n\\x1a\\n\\x00\\x00\\x00\\rIHDR\\x00\\x00\\x00\\x01\\x00\\x00\\x00\\x01\\x08\\x06\\x00\\x00\\x00\\x1f\\x15\\xc4\\x89\\x00\\x00\\x00\\nIDATx\\x9cc\\x00\\x01\\x00\\x00\\x05\\x00\\x01\\r\\n\\x2e\\xae\\x00\\x00\\x00\\x00IEND\\xaeB`\\x82')
        elif self.path == "/slow-resource-1":
            time.sleep(3)  # Simulate slow resource
            self.send_response(200)
            self.send_header("Content-type", "text/plain")
            self.end_headers()
            self.wfile.write(b"Slow resource 1 content")
        elif self.path == "/slow-resource-2":
            time.sleep(1)  # Simulate another slow resource
            self.send_response(200)
            self.send_header("Content-type", "text/plain")
            self.end_headers()
            self.wfile.write(b"Slow resource 2 content")
        else:
            self.send_error(404)

class ThreadedTCPServer(socketserver.ThreadingMixIn, socketserver.TCPServer):
    pass

def run_server(port):
    # Use allow_reuse_address to avoid "Address already in use" errors on retry
    socketserver.TCPServer.allow_reuse_address = True
    with ThreadedTCPServer(("127.0.0.1", port), SlowHandler) as httpd:
        httpd.serve_forever()

def make_request(path, method="GET", data=None):
    url = f"{BASE_URL}{path}"
    req = urllib.request.Request(url, method=method)
    if data:
        data = json.dumps(data).encode()
        req.add_header("Content-Type", "application/json")
    
    try:
        with urllib.request.urlopen(req, data=data, timeout=15) as res:
            return res.status, json.loads(res.read().decode())
    except Exception as e:
        return None, str(e)

def main():
    port = 8889 # Use a different port just in case
    server_thread = threading.Thread(target=run_server, args=(port,), daemon=True)
    server_thread.start()
    
    test_url = f"http://127.0.0.1:{port}/"
    print(f"Testing Network Idle with {test_url}")
    
    # 1. Create instance
    print("Creating instance...")
    status, body = make_request("/scraper/instances", "POST")
    if status != 200:
        print(f"Failed to create instance: {body}")
        return
    instance_id = body["instanceId"]
    print(f"Created instance: {instance_id}")
    
    try:
        # 2. Navigate
        print("Navigating...")
        nav_status, nav_body = make_request(f"/scraper/instances/{instance_id}/navigate", "POST", {"url": test_url})
        if nav_status != 200:
            print(f"Navigation failed: {nav_body}")
            return
        
        # 3. Wait for Network Idle
        # The page will be idle after 1s (delay) + 2s (fetch) = 3s.
        # We call waitForNetworkIdle immediately.
        print("Calling waitForNetworkIdle (should take ~3s)...")
        start_time = time.time()
        status, body = make_request(f"/scraper/instances/{instance_id}/waitForNetworkIdle", "POST", {"timeout": 10000})
        end_time = time.time()
        
        duration = end_time - start_time
        print(f"waitForNetworkIdle returned after {duration:.2f}s")
        print(f"Result: {body}")
        
        # 0.5s idle threshold + 3s max delay = 3.5s total expected
        if body.get("ok") and 2.5 <= duration <= 6.0:
            print("SUCCESS: waitForNetworkIdle correctly waited for background network activity.")
        else:
            print(f"FAILURE: waitForNetworkIdle took {duration:.2f}s, which is outside expected range (2.5-6.0s).")

        # 4. Immediate call when idle
        print("Calling waitForNetworkIdle again (should be immediate)...")
        start_time = time.time()
        status, body = make_request(f"/scraper/instances/{instance_id}/waitForNetworkIdle", "POST", {"timeout": 10000})
        end_time = time.time()
        duration = end_time - start_time
        print(f"waitForNetworkIdle returned after {duration:.2f}s")
        if body.get("ok") and duration < 1.0:
            print("SUCCESS: waitForNetworkIdle returned quickly when already idle.")
        else:
            print("FAILURE: waitForNetworkIdle took too long when already idle.")

    finally:
        print("Destroying instance...")
        make_request(f"/scraper/instances/{instance_id}", "DELETE")

if __name__ == "__main__":
    main()

File: tools/os-test/verify_network_idle_rigorous.ts
#!/usr/bin/env -S deno run -A
// SPDX-License-Identifier: MPL-2.0

import { parseArgs } from "@std/cli/parse-args";

const DEFAULT_BASE_URL = "http://127.0.0.1:58261";

const TEST_PAGE_HTML = `<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8" />
  <title>Network Idle Test</title>
  <link rel="stylesheet" href="/slow-css" />
</head>
<body>
  <h1 id="title">Network Idle Test</h1>
  <img src="/slow-image" />
  <div id="logs"></div>
  <script>
    Promise.all([
      fetch('/slow-resource-1'),
      fetch('/slow-resource-2'),
    ]).then(() => {
      const done = document.createElement('p');
      done.id = 'done';
      done.textContent = 'done';
      document.body.appendChild(done);
    });
  </script>
</body>
</html>`;

type RequestResult = {
  status: number | null;
  body: unknown;
};

import { sleep } from "../src/async_utils.ts";

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

async function makeRequest(
  baseUrl: string,
  path: string,
  options: {
    method?: "GET" | "POST" | "DELETE";
    data?: unknown;
    timeoutMs?: number;
  } = {},
): Promise<RequestResult> {
  const url = `${baseUrl}${path}`;
  const timeoutMs = options.timeoutMs ?? 15_000;

  try {
    const response = await fetch(url, {
      method: options.method ?? "GET",
      headers:
        options.data !== undefined
          ? { "content-type": "application/json" }
          : undefined,
      body:
        options.data !== undefined ? JSON.stringify(options.data) : undefined,
      signal: AbortSignal.timeout(timeoutMs),
    });

    const raw = await response.text();
    let body: unknown = raw;
    if (raw.length > 0) {
      try {
        body = JSON.parse(raw);
      } catch {
        body = raw;
      }
    }

    return { status: response.status, body };
  } catch (error) {
    const detail = error instanceof Error ? error.message : String(error);
    return { status: null, body: detail };
  }
}

function startSlowServer(port: number): {
  url: string;
  close: () => Promise<void>;
} {
  const server = Deno.serve(
    { hostname: "127.0.0.1", port },
    async (request) => {
      const url = new URL(request.url);

      if (url.pathname === "/") {
        return new Response(TEST_PAGE_HTML, {
          headers: { "content-type": "text/html; charset=utf-8" },
        });
      }

      if (url.pathname === "/slow-css") {
        await sleep(600);
        return new Response("body { color: #222; }", {
          headers: { "content-type": "text/css" },
        });
      }

      if (url.pathname === "/slow-image") {
        await sleep(900);
        const png = Uint8Array.from([
          137, 80, 78, 71, 13, 10, 26, 10, 0, 0, 0, 13, 73, 72, 68, 82, 0, 0, 0,
          1, 0, 0, 0, 1, 8, 6, 0, 0, 0, 31, 21, 196, 137, 0, 0, 0, 10, 73, 68,
          65, 84, 120, 156, 99, 0, 1, 0, 0, 5, 0, 1, 13, 10, 46, 174, 0, 0, 0,
          0, 73, 69, 78, 68, 174, 66, 96, 130,
        ]);
        return new Response(png, {
          headers: { "content-type": "image/png" },
        });
      }

      if (url.pathname === "/slow-resource-1") {
        await sleep(1200);
        return new Response("Slow resource 1", {
          headers: { "content-type": "text/plain" },
        });
      }

      if (url.pathname === "/slow-resource-2") {
        await sleep(700);
        return new Response("Slow resource 2", {
          headers: { "content-type": "text/plain" },
        });
      }

      return new Response("Not Found", { status: 404 });
    },
  );

  const addr = server.addr as Deno.NetAddr;
  return {
    url: `http://127.0.0.1:${addr.port}/`,
    close: async () => {
      await server.shutdown();
    },
  };
}

export async function main(argv: string[] = Deno.args): Promise<number> {
  const parsed = parseArgs(argv, {
    string: ["base-url", "port", "timeout"],
    default: {
      "base-url": DEFAULT_BASE_URL,
      port: "0",
      timeout: "10000",
    },
  });

  const baseUrl = String(parsed["base-url"]);
  const serverPort = Number(parsed.port);
  const networkIdleTimeout = Number(parsed.timeout);

  let failures = 0;
  const assertCase = (name: string, condition: boolean, detail = "") => {
    if (condition) {
      console.log(`[OK] ${name}`);
      return;
    }
    failures += 1;
    console.log(`[FAIL] ${name}${detail ? ` - ${detail}` : ""}`);
  };

  const localServer = await startSlowServer(serverPort);
  const testUrl = localServer.url;
  console.log(`Testing network idle with ${testUrl}`);

  let instanceId: string | null = null;

  try {
    const created = await makeRequest(baseUrl, "/scraper/instances", {
      method: "POST",
    });
    assertCase(
      "Create scraper instance",
      created.status === 200 &&
        isRecord(created.body) &&
        typeof created.body.instanceId === "string",
      String(created.body),
    );

    if (
      !(
        created.status === 200 &&
        isRecord(created.body) &&
        typeof created.body.instanceId === "string"
      )
    ) {
      return 1;
    }

    instanceId = created.body.instanceId;

    const navigated = await makeRequest(
      baseUrl,
      `/scraper/instances/${instanceId}/navigate`,
      {
        method: "POST",
        data: { url: testUrl },
      },
    );
    assertCase(
      "Navigate scraper",
      navigated.status === 200,
      String(navigated.body),
    );
    if (navigated.status !== 200) {
      return 1;
    }

    const waitDone = await makeRequest(
      baseUrl,
      `/scraper/instances/${instanceId}/waitForElement`,
      {
        method: "POST",
        data: { selector: "#done", timeout: 10000 },
        timeoutMs: 20_000,
      },
    );
    assertCase(
      "waitForElement(#done) contract",
      waitDone.status === 200 &&
        isRecord(waitDone.body) &&
        waitDone.body.ok === true &&
        waitDone.body.found === true,
      String(waitDone.body),
    );

    const firstStart = performance.now();
    const firstIdle = await makeRequest(
      baseUrl,
      `/scraper/instances/${instanceId}/waitForNetworkIdle`,
      {
        method: "POST",
        data: { timeout: networkIdleTimeout },
        timeoutMs: 20_000,
      },
    );
    const firstDuration = (performance.now() - firstStart) / 1000;
    console.log(
      `First waitForNetworkIdle duration: ${firstDuration.toFixed(2)}s`,
    );

    assertCase(
      "waitForNetworkIdle contract (first)",
      firstIdle.status === 200 &&
        isRecord(firstIdle.body) &&
        typeof firstIdle.body.ok === "boolean",
      String(firstIdle.body),
    );

    assertCase(
      "waitForNetworkIdle should succeed after done",
      isRecord(firstIdle.body) && firstIdle.body.ok === true,
      String(firstIdle.body),
    );

    const secondStart = performance.now();
    const secondIdle = await makeRequest(
      baseUrl,
      `/scraper/instances/${instanceId}/waitForNetworkIdle`,
      {
        method: "POST",
        data: { timeout: networkIdleTimeout },
        timeoutMs: 20_000,
      },
    );
    const secondDuration = (performance.now() - secondStart) / 1000;
    console.log(
      `Second waitForNetworkIdle duration: ${secondDuration.toFixed(2)}s`,
    );

    assertCase(
      "waitForNetworkIdle contract (second)",
      secondIdle.status === 200 &&
        isRecord(secondIdle.body) &&
        typeof secondIdle.body.ok === "boolean",
      String(secondIdle.body),
    );

    const secondCeiling = Math.max(
      firstDuration + 6.0,
      networkIdleTimeout / 1000 + 2.0,
      12.0,
    );
    assertCase(
      "Second wait should not be dramatically slower",
      secondDuration <= secondCeiling,
      `first=${firstDuration.toFixed(2)}s second=${secondDuration.toFixed(2)}s`,
    );

    const shortIdle = await makeRequest(
      baseUrl,
      `/scraper/instances/${instanceId}/waitForNetworkIdle`,
      {
        method: "POST",
        data: { timeout: 10 },
        timeoutMs: 20_000,
      },
    );

    assertCase(
      "waitForNetworkIdle short-timeout contract",
      shortIdle.status === 200 &&
        isRecord(shortIdle.body) &&
        typeof shortIdle.body.ok === "boolean",
      String(shortIdle.body),
    );
  } finally {
    if (instanceId) {
      await makeRequest(baseUrl, `/scraper/instances/${instanceId}`, {
        method: "DELETE",
      });
    }
    await localServer.close();
  }

  console.log(`Completed with ${failures} failure(s)`);
  return failures > 0 ? 1 : 0;
}

if (import.meta.main) {
  const code = await main(Deno.args);
  Deno.exit(code);
}
